src/modules/fancyzones/lib/FancyZonesDataTypes.cpp

```diff
@@ -112,4 +114,109 @@ namespace FancyZonesDataTypes
             cellRow.resize(m_columns, 0);
         }
     }
+
+    bool DeviceIdData::empty() const
+    {
+        return *this == DeviceIdData{};
+    }
+
+    std::wstring DeviceIdData::Serialize() const
+    {
+        static std::wstring vdId = L"{00000000-0000-0000-0000-000000000000}";
+        wil::unique_cotaskmem_string virtualDesktopIdStr;
+        if (SUCCEEDED(StringFromCLSID(virtualDesktopId, &virtualDesktopIdStr)))
+        {
+            vdId = std::wstring{ virtualDesktopIdStr.get() };
+        }
+
+        return deviceName + L"_" + std::to_wstring(width) + L"_" + std::to_wstring(height) + L"_" + vdId;
+    }
+
+    DeviceIdData DeviceIdData::Parse(const std::wstring& deviceId)
+    {
+        DeviceIdData data;
+
+        std::wstring temp;
+        std::wstringstream wss(deviceId);
+
+        /*
+        Important fix for device info that contains a '_' in the name:
+        1. first search for '#'
+        2. Then split the remaining string by '_'
+        */
+
+        // Step 1: parse the name until the #, then to the '_'
+        if (deviceId.find(L'#') != std::string::npos)
+        {
+            std::getline(wss, temp, L'#');
+
+            data.deviceName = temp;
+
+            if (!std::getline(wss, temp, L'_'))
+            {
+                return {};
+            }
+
+            data.deviceName += L"#" + temp;
+        }
+        else if (std::getline(wss, temp, L'_') && !temp.empty())
+        {
+            data.deviceName = temp;
+        }
+        else
+        {
+            return {};
+        }
+
+        // Step 2: parse the rest of the id
+        std::vector<std::wstring> parts;
+        while (std::getline(wss, temp, L'_'))
+        {
+            parts.push_back(temp);
+        }
+
+        if (parts.size() != 3 && parts.size() != 4)
+        {
+            return {};
+        }
+
+        /*
+        Refer to ZoneWindowUtils::GenerateUniqueId parts contain:
```

Comment:
This function is called `FancyZonesUtils::GenerateUniqueId`, so
let's update the comment.

Reply:
Updated. Thanks